### PR TITLE
Speed improvements

### PR DESCRIPTION
**Pull request by MauriceHendrix**
## Description

Changed when optimisations in equations are done. The log and pow related optimisations are only done for equations that contain logs or pows the Mul optimisations (removing 1.0 * ) is moved to the printer. Optimisations in the equations in get_equations_for_ionic_vars are and _convert_membrane_stimulus_current no longer applied.

I also realised that empty lists of state vars and jacobians ae possible in backwards Euler models and the code deals just fine with that so removed the check preventing that.

## Motivation and Context

Applying all optimisations to all equations was slowing things down. Also the equations in get_equations_for_ionic_vars is later re-processed in _get_extended_equations_for_ionic_vars and optimisation happens there. The call to get_equations_for in _convert_membrane_stimulus_current is only used to calculate the sign for GetIntracellularAreaStimulus, the optimisations have no effect on the output generated there so are not needed.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

Speed improvement

## Documentation checklist

- [ ] I have updated all documentation in the code where necessary.
- [ ] I have checked spelling in all (new) comments and documentation.
- [ ] I have added a note to RELEASE.md if relevant (new feature, breaking change, or notable bug fix).

## Testing
- [X] Testing is done automatically and codecov shows test coverage
- [ ] This cannot be tested automatically 

All models are tested with chaste and working. WiIll update all models once PR is approved.

### CODE REVIEW

File: chaste_codegen/chaste_model.py [hunk 1]
```diff
@@ -11,15 +11,14 @@
...
     Pow,
     Wild,
     log,
     simplify,
     sympify,
 )
-from sympy.codegen.cfunctions import log10
-from sympy.codegen.rewriting import ReplaceOptim, optimize, optims_c99
+from sympy.codegen.cfunctions import log2, log10
+from sympy.codegen.rewriting import ReplaceOptim, optimize
```
Comment (jonc125) [inline chaste_codegen/chaste_model.py:21]: What does `optims_c99` do that we're now missing out?

File: chaste_codegen/chaste_model.py [hunk 2]
```diff
@@ -350,7 +359,7 @@ def _convert_membrane_stimulus_current(self):
         if self._membrane_stimulus_current_orig is None:
             return None, self._stimulus_units
         # get derivative equations
-        d_eqs = self.get_equations_for(self._y_derivatives)
+        d_eqs = self._model.get_equations_for(self._y_derivatives)
```
Comment (jonc125) [inline chaste_codegen/chaste_model.py:362]: I'd be inclined to add an `optimise=False` argument to `self.get_equations_for` as otherwise it isn't obvious in the call what effect the different method has.

File: chaste_codegen/data/tests/chaste_reference_models/CVODE_DATA_CLAMP/Shannon2004_with_modifiers.cpp [hunk 3]
```diff
@@ -306,7 +306,7 @@
...
-        const double var_IK1_K1_gate__beta_K1 = (0.49124000000000001 * exp(0.43983232 + 0.080320000000000003 * var_chaste_interface__cell__V - 0.080320000000000003 * var_reversal_potentials__E_K) + exp(-36.698642499999998 + 0.061749999999999999 * var_chaste_interface__cell__V - 0.061749999999999999 * var_reversal_potentials__E_K)) / (1.0 + exp(-2.4444678999999998 + 0.51429999999999998 * var_reversal_potentials__E_K - 0.51429999999999998 * var_chaste_interface__cell__V)); // per_millisecond
+        const double var_IK1_K1_gate__beta_K1 = (exp(-36.698642499999998 + 0.061749999999999999 * var_chaste_interface__cell__V - 0.061749999999999999 * var_reversal_potentials__E_K) + 0.49124000000000001 * exp(0.43983232 + 0.080320000000000003 * var_chaste_interface__cell__V - 0.080320000000000003 * var_reversal_potentials__E_K)) / (1.0 + exp(-2.4444678999999998 + 0.51429999999999998 * var_reversal_potentials__E_K - 0.51429999999999998 * var_chaste_interface__cell__V)); // per_millisecond
```
Comment (jonc125) [inline chaste_codegen/data/tests/chaste_reference_models/CVODE_DATA_CLAMP/Shannon2004_with_modifiers.cpp:309]: OK, this is just a reordering of terms, presumably because the `Mul` handling has moved?

Comment (jonc125) [reply chaste_codegen/data/tests/chaste_reference_models/CVODE_DATA_CLAMP/Shannon2004_with_modifiers.cpp:309]: It's actually going back to what it was before we added that 'optimisation' I think!

File: chaste_codegen/data/tests/chaste_reference_models/Cvode_with_jacobian/hodgkin_huxley_squid_axon_model_1952_modified.cpp [hunk 4]
```diff
@@ -178,7 +178,7 @@
         IJth(rJacobian, 0, 1) = mSetVoltageDerivativeToZero ? 0.0 : (-360.0 * var_x1 * pow(var_chaste_interface__sodium_channel_m_gate__m, 2) * var_chaste_interface__sodium_channel_h_gate__h);
         IJth(rJacobian, 1, 1) = -4.0 * var_x2 + var_x7 * var_x8;
         IJth(rJacobian, 0, 2) = mSetVoltageDerivativeToZero ? 0.0 : (-var_x0 * var_x1);
-        IJth(rJacobian, 2, 2) = -1 / var_x11 - 0.070000000000000007 * var_x9;
+        IJth(rJacobian, 2, 2) = -1.0 / var_x11 - 0.070000000000000007 * var_x9;
```
Comment (jonc125) [inline chaste_codegen/data/tests/chaste_reference_models/Cvode_with_jacobian/hodgkin_huxley_squid_axon_model_1952_modified.cpp:181]: Any idea why this change has happened?

File: chaste_codegen/data/tests/chaste_reference_models/Cvode_with_jacobian/dynamic_Shannon2004.cpp [hunk 5]
```diff
@@ -997,12 +997,12 @@
...
-        const double var_x242 = 1 / var_x241;
-        const double var_x243 = exp(0.24440000000000001 * var_chaste_interface__cell__V);
-        const double var_x244 = exp(-0.043909999999999998 * var_chaste_interface__cell__V);
-        const double var_x245 = -127140.0 * var_x243 - 3.4740000000000003e-5 * var_x244;
+        const double var_x240 = exp(0.24440000000000001 * var_chaste_interface__cell__V);
+        const double var_x241 = exp(-0.043909999999999998 * var_chaste_interface__cell__V);
+        const double var_x242 = -127140.0 * var_x240 - 3.4740000000000003e-5 * var_x241;
+        const double var_x243 = exp(24.640530000000002 + 0.311 * var_chaste_interface__cell__V);
+        const double var_x244 = 1.0 + var_x243;
+        const double var_x245 = 1 / var_x244;
```
Comment (jonc125) [inline chaste_codegen/data/tests/chaste_reference_models/Cvode_with_jacobian/dynamic_Shannon2004.cpp:1005]: Again just a reordering, but harder to spot that here!

File: chaste_codegen/data/tests/chaste_reference_models/Cvode_with_jacobian/dynamic_Shannon2004.cpp [hunk 6]
```diff
@@ -1062,98 +1062,99 @@
...
-        const double var_x369 = 1205.9369202226344 * NV_Ith_S(mParameters, 0);
-        const double var_x370 = 64.935064935064929 * var_x360;
-        const double var_x371 = 100.0 * var_chaste_interface__Ca_buffer__Ca_SLB_jct;
-        const double var_x372 = 100.0 * var_chaste_interface__Ca_buffer__Ca_SLHigh_jct;
-        const double var_x373 = 100.0 * var_chaste_interface__Ca_buffer__Ca_jct;
-        const double var_x374 = -var_x366;
-        const double var_x375 = -var_x373;
-        const double var_x376 = -var_x230;
-        const double var_x377 = -var_x232;
+        const double var_x305 = var_x304;
```
Comment (jonc125) [inline chaste_codegen/data/tests/chaste_reference_models/Cvode_with_jacobian/dynamic_Shannon2004.cpp:1065]: Boggle! Why a different name for the same term?

File: chaste_codegen/data/tests/chaste_reference_models/GRL1Opt/dynamic_matsuoka_model_2003.cpp [hunk 7]
```diff
@@ -125,7 +125,7 @@
...
-        const double var_chaste_interface__i_ionic = 1.0 * HeartConfig::Instance()->GetCapacitance() * var_membrane__i_tot / mParameters[0]; // uA_per_cm2
+        const double var_chaste_interface__i_ionic = HeartConfig::Instance()->GetCapacitance() * var_membrane__i_tot / mParameters[0]; // uA_per_cm2
```
Comment (jonc125) [inline chaste_codegen/data/tests/chaste_reference_models/GRL1Opt/dynamic_matsuoka_model_2003.cpp:128]: Looks like this was missed out of an earlier regeneration?

File: chaste_codegen/data/tests/chaste_reference_models/GRL1Opt/dynamic_matsuoka_model_2003.cpp [hunk 8]
```diff
@@ -1539,7 +1539,7 @@
             const double var_x108 = var_x61 * var_x66;
             const double var_x110 = var_x62 * var_x79;
             const double var_x111 = var_x110 * var_x65;
-            const double var_x113 = 1.0 * var_x73;
+            const double var_x113 = var_x73;
```
Comment (jonc125) [inline chaste_codegen/data/tests/chaste_reference_models/GRL1Opt/dynamic_matsuoka_model_2003.cpp:1542]: Ah, I see why the bizarre effect happens! Having the 1.0* removed just when printing means the assignment isn't elided in the Jacobian calculation.

File: chaste_codegen/data/tests/chaste_reference_models/RLopt/dynamic_bondarenko_model_2004_apex.cpp [hunk 9]
```diff
@@ -102,9 +102,9 @@
...
         const double var_slow_delayed_rectifier_potassium_current__i_Ks = 0.0057499999999999999 * pow(var_chaste_interface__slow_delayed_rectifier_potassium_current__nKs, 2) * (-var_fast_transient_outward_potassium_current__E_K + var_chaste_interface__membrane__V); // picoA_per_picoF
         const double var_slow_transient_outward_potassium_current__i_Kto_s = 0; // picoA_per_picoF
-        const double var_fast_sodium_current__E_Na = 25.674321243523316 * log(140600.0 / (0.1111111111111111 * var_chaste_interface__potassium_concentration__Ki + var_chaste_interface__sodium_concentration__Nai)); // millivolt
+        const double var_fast_sodium_current__E_Na = 25.674321243523316 * log(126540.0 / (0.10000000000000001 * var_chaste_interface__potassium_concentration__Ki + 0.90000000000000002 * var_chaste_interface__sodium_concentration__Nai)); // millivolt
```
Comment (jonc125) [inline chaste_codegen/data/tests/chaste_reference_models/RLopt/dynamic_bondarenko_model_2004_apex.cpp:105]: Why have these changed?

File: chaste_codegen/chaste_model.py (hunk 1 above)
Comment (MauriceHendrix) [reply chaste_codegen/chaste_model.py:21]: optims_c99 = (expm1_opt, log1p_opt, exp2_opt, log2_opt, log2const_opt)

Comment (jonc125) [reply chaste_codegen/chaste_model.py:21]: Worth getting some data for each of these as to how often they can be applied in our full suite of models, and how expensive they are to apply, to guide which ones we decide are worth doing.

File: chaste_codegen/data/tests/chaste_reference_models/CVODE_DATA_CLAMP/Shannon2004_with_modifiers.cpp (hunk 3 above)
Comment (MauriceHendrix) [reply chaste_codegen/data/tests/chaste_reference_models/CVODE_DATA_CLAMP/Shannon2004_with_modifiers.cpp:309]: yup

File: chaste_codegen/data/tests/chaste_reference_models/GRL1Opt/dynamic_matsuoka_model_2003.cpp (hunk 7 above)
Comment (MauriceHendrix) [reply chaste_codegen/data/tests/chaste_reference_models/GRL1Opt/dynamic_matsuoka_model_2003.cpp:128]: yes as that equation is added after the get_equations_for call

File: chaste_codegen/chaste_model.py (hunk 1 above)
Comment (MauriceHendrix) [reply chaste_codegen/chaste_model.py:21]: optims_c99 = (expm1_opt, log1p_opt, exp2_opt, log2_opt, log2const_opt)

expm1_opt -  exp(x-1) --> expm1(x) -- might want it back I'll check if it influences the output
log1p_opt - log(x+1) --> log1p(x) -- might want it back I'll check if it influences the output
exp2_opt - Pow(x, 2) --> exp2(x) -- might want it back I'll check if it influences the output
log2_opt - log(x)/log(2) -> log2(x) -- doesn't make a difference to the output
log2const_opt - log(2)*log2(_w) -> log(_w) -- doesn't make a difference to the output

File: chaste_codegen/data/tests/chaste_reference_models/GRL1Opt/dynamic_matsuoka_model_2003.cpp (hunk 8 above)
Comment (MauriceHendrix) [reply chaste_codegen/data/tests/chaste_reference_models/GRL1Opt/dynamic_matsuoka_model_2003.cpp:1542]: yes exactly, that's a downside, however I think it's an acceptable compromise

File: chaste_codegen/chaste_model.py (hunk 2 above)
Comment (MauriceHendrix) [reply chaste_codegen/chaste_model.py:356]: I did think of that but wasn't sure it was worth it

File: chaste_codegen/data/tests/chaste_reference_models/RLopt/dynamic_bondarenko_model_2004_apex.cpp (hunk 9 above)
Comment (MauriceHendrix) [reply chaste_codegen/data/tests/chaste_reference_models/RLopt/dynamic_bondarenko_model_2004_apex.cpp]: Seems to be the version in the master was a side effect of the 1.0* optimisation

**Comment by MauriceHendrix** [comment]
@jonc125 I have just pushed a bunch of further improvements. This is mainly:

- removing a few unnecessary calls
- as much as possible intermediate sets of variables or equations are now handled as sets rather than lists since testing membership of sets is faster.

At the moment the thing that definately takes the most time is `cellmlmanip.load_model`.
